File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.py
def solution(n):
    arr = [[0] * n for _ in range(n)]
    x, y = 0, 0
    dx = [0, 1, 0, -1]
    dy = [1, 0, -1, 0]
    direction = 0
    num = 1

    while num <= n**2:
        arr[x][y] = num
        num += 1
        nx, ny = x + dx[direction], y + dy[direction]
        if 0 <= nx < n and 0 <= ny < n and arr[nx][ny] == 0:
            x, y = nx, ny
        else:
            direction = (direction + 1) % 4
            x, y = x + dx[direction], y + dy[direction]

    return arr

File: 프로그래머스/0/181832. 정수를 나선형으로 배치하기/정수를 나선형으로 배치하기.js
function solution(n) {
    let answer = new Array(n).fill(null).map(() => new Array(n).fill(0));
    
    let dx = [0, 1, 0, -1];
    let dy = [1, 0, -1, 0];
    let direction = 0;
    let x = 0, y = 0; 
    
    for (let num = 1; num <= n ** 2; num++) {
        answer[x][y] = num;
        
        let nx = x + dx[direction];
        let ny = y + dy[direction];
        
        if (0 <= nx && nx < n && 0 <= ny && ny < n && answer[nx][ny] === 0){           x = nx;
            y = ny;
        } else {
            direction = (direction + 1) % 4;
            x += dx[direction];
            y += dy[direction];
        }
    }
    
    return answer;
}
